### PR TITLE
Add new database backend to the header cache: TDB

### PR DESCRIPTION
**Pull request by mcmilk**
* **What does this PR do?**

TDB stands for "Trivial database" - most people may know it from samba ;-)
This backend is fast and gives good results together with the new compression functions.

### CODE REVIEW

File: auto.def [hunk 1]
```diff
@@ -1017,6 +1019,22 @@ if {[get-define want-qdbm]} {
...
 }
 
+###############################################################################
+# Header Cache - TDB
+if {[get-define want-tdb]} {
+  if {[get-define want-pkgconf]} {
+    pkgconf true tdb
+    define-feature tdb
+  } else {
+    if {![check-inc-and-lib tc [opt-val with-tdb $prefix] \
```
Comment (gahr) [inline auto.def]: s/tc/tdb/

File: hcache/tdb.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @page hc_tdb TDB
+ *
+ * Use a GNU dbm file as a header cache backend.
```
Comment (gahr) [inline hcache/tdb.c]: Needs fixing

File: hcache/tdb.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+/**
+ * hcache_tdb_store - Implements HcacheOps::store()
+ */
+static int hcache_tdb_store(void *ctx, const char *key, size_t keylen, void *data, size_t dlen)
+{
+  if (!ctx)
+    return -1;
+
+  TDB_DATA dkey;
+  TDB_DATA databuf;
```
Comment (gahr) [inline hcache/tdb.c]: Elsewhere you call this just ‘data’

File: hcache/tdb.c [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+#include <fcntl.h>
+#include <tdb.h>
+
+/**
+ * hcache_tdb_open - Implements HcacheOps::open()
+ */
+static void *hcache_tdb_open(const char *path)
+{
+  int flags = TDB_NOLOCK|TDB_INCOMPATIBLE_HASH|TDB_NOSYNC;
+  return tdb_open(path, 33657, flags, O_CREAT|O_RDWR, 00600);
```
Comment (gahr) [inline hcache/tdb.c]: Can you provide some background on te choice of flags (especially the undocumented TDB_INCOMPATIBLE_HASH) and of hash size? Perhaps as a comment in the code.

File: hcache/tdb.c (hunk 3 above)
Comment (mcmilk) [reply hcache/tdb.c]: data is the parameter of that function :/

Comment (gahr) [reply hcache/tdb.c]: Oh, sure... nevermind

File: hcache/tdb.c (hunk 4 above)
Comment (mcmilk) [reply hcache/tdb.c]: - TDB_INCOMPATIBLE_HASH is a lot faster, and we have no other TDB's currently
- the number was some try+fail ... until the speed on my test was nice

**Comment by mcmilk** [comment]
I have created an overview of the timings (sortable) and a shell script for the benchmarking here: https://mcmilk.de/code/2020/

**Comment by flatcap** [comment]
Rebased, clang-format, light tidy

**Comment by gahr** [comment]
Your benchmark looks nice. Would you be willing to work on making https://github.com/neomutt/neomutt/tree/master/contrib/hcache-bench better?

**Comment by mcmilk** [comment]
> Your benchmark looks nice. Would you be willing to work on making https://github.com/neomutt/neomutt/tree/master/contrib/hcache-bench better?

Yes, I will do so.

**Pull request merged**